storage/src/main/java/tech/pegasys/artemis/storage/ChainStorageServer.java

```diff
@@ -81,15 +83,18 @@ public void onStoreRequest(final GetStoreRequest request) {
     eventBus.post(new GetStoreResponse(request.getId(), getStore()));
   }
 
-  @Subscribe
-  public void onStoreDiskUpdate(final StoreDiskUpdateEvent event) {
-    final DatabaseUpdateResult result = database.update(event);
-    handleStoreUpdate(result);
-    eventBus.post(new StoreDiskUpdateCompleteEvent(event.getTransactionId(), result));
+  @Override
+  public SafeFuture<DiskUpdateResult> onDiskUpdate(final DiskUpdate event) {
+    return SafeFuture.supplyAsync(
```

Comment:
We shouldn't use `supplyAsync` here or events will be recorded in parallel and potentially out of order.  Just `SafeFuture.of`.

Reply:
Actually now I think about it, as a follow up to this we should be able to remove the `.join()` calls after each `commit()` because we know each change will be applied in order.

Though there is still a question of what to do if the returned future throws an `Exception`. Will need some thought, but would be really nice if we can let our processing run ahead of storage.

Reply:
Makes sense. Done.

Reply:
Thinking out loud, maybe, in a scenario where the disk is throwing exceptions, we can start logging warning level messages, but keep executing since being able to validate is more important than writing to disk. Through the messages, the user would know something needs fixing, but still not lose their stake in the meanwhile.

Reply:
https://pegasys1.atlassian.net/secure/RapidBoard.jspa?rapidView=67&modal=detail&selectedIssue=BC-319